lib/Runtime/Language/JavascriptOperators.cpp

```diff
@@ -2152,8 +2152,8 @@ namespace Js
                 {
                     Assert((flags & Data) == Data && (flags & Writable) == None);
 
-                    JavascriptError::ThrowCantAssignIfStrictMode(propertyOperationFlags, requestContext);
                     requestContext->GetThreadContext()->AddImplicitCallFlags(ImplicitCall_NoOpSet);
+                    JavascriptError::ThrowCantAssignIfStrictMode(propertyOperationFlags, requestContext);
```

Comment:
I think bailing out (and then throwing from the interpreter if appropriate) is actually preferable, because it gives us a chance to rejit with the knowledge that the function will trigger implicit calls. That won’t keep us from throwing in future, but it’ll keep us from optimizing in a way that requires us to bail out.

--Paul

From: Michael Ferris [mailto:notifications@github.com]
Sent: Thursday, April 14, 2016 2:40 PM
To: Microsoft/ChakraCore <ChakraCore@noreply.github.com>
Cc: Paul Leathers <pleath@microsoft.com>
Subject: Re: [Microsoft/ChakraCore] Disable implicit exceptions on BailOutOnImplicitCallsPreOp (#806)


In lib/Runtime/Language/JavascriptOperators.cpp<https://github.com/Microsoft/ChakraCore/pull/806#discussion_r59796631>:

>                      requestContext->GetThreadContext()->AddImplicitCallFlags(ImplicitCall_NoOpSet);

> +                    JavascriptError::ThrowCantAssignIfStrictMode(propertyOperationFlags, requestContext);

It is a little confusing how this case should be handled.
The implicit call possibly already happened and we want to throw/bailout.
With my change, in this scenario it means that if we disabled implicit calls we won't throw and bailout instead.
The question is, should we prioritize bailing out or throwing ?

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/ChakraCore/pull/806/files/6dc49f470b757ae16ed6dc64ca1d9df0eb70cf85#r59796631>